pydatastructs/linear_data_structures/tests/test_linked_lists.py

```diff
@@ -10,18 +10,19 @@ def test_DoublyLinkedList():
     dll.append(1)
     dll.append_left(2)
     dll.append(3)
-    dll.insert_after(dll[1], 4)
-    dll.insert_after(dll[-1], 6)
-    dll.insert_before(dll[0], 1)
+    dll.insert_after(dll[-1], 4)
+    dll.insert_after(dll[2], 6)
+    dll.insert_before(dll[4], 1)
+    dll.insert_before(dll[0], 7)
     dll.insert_at(0, 2)
     dll.insert_at(-1, 9)
     dll.extract(2)
     dll.extract(0)
     dll.extract(-1)
     dll[-2].data = 0
-    assert str(dll) == "[1, 5, 4, 1, 0, 9]"
-    assert len(dll) == 6
-    assert raises(IndexError, lambda: dll.insert_at(7, None))
+    assert str(dll) == "[7, 5, 1, 6, 1, 0, 9]"
```

Comment:
IMO, 9(i.e., dll[-1]) shouldn't have been there. `dll.extract(-1)` should have removed that.

Reply:
I will open a separate issue for this. Let's merge the PR for now.

Reply:
`dll.insert_at(-1,9)` inserts 9 at -1 locations i.e. at index 7. So at that time last element is 4 not 9.

dll = [2, 7, 2, 5, 1, 6, 1, 3, 4]
`dll.insert_at(-1,9)`
dll = [2, 7, 2, 5, 1, 6, 1, 3, 9, 4]

Reply:
Ah! Right -1 translated to the positive index for the previous list is index 8. So yeah, not a bug.